pkg/api/customization/pipeline/utils.go

```diff
@@ -0,0 +1,58 @@
+package pipeline
+
+import (
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/remote"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/satori/uuid"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func refreshReposByID(sourceCodeCredentials v3.SourceCodeCredentialInterface, sourceCodeRepositories v3.SourceCodeRepositoryInterface, sourceCodeCredentialID string) ([]v3.SourceCodeRepository, error) {
+
+	credential, err := sourceCodeCredentials.Get(sourceCodeCredentialID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	remoteType := credential.Spec.SourceCodeType
+
+	mockConfig := v3.ClusterPipeline{
+		Spec: v3.ClusterPipelineSpec{
+			GithubConfig: &v3.GithubClusterConfig{},
+		},
+	}
+	remote, err := remote.New(mockConfig, remoteType)
+	if err != nil {
+		return nil, err
+	}
+	repos, err := remote.Repos(credential)
+	if err != nil {
+		return nil, err
+	}
+
+	//remove old repos
+	repoList, err := sourceCodeRepositories.List(metav1.ListOptions{})
```

Comment:
use cache？

Reply:
The method is invoked in handlers and we cannot fetch from cache there.